### ISSUE TITLE
error: failed to run custom build command for `openssl-sys v0.9.58`: Cross compilation

### ISSUE THREAD

**Issue by huguesBouvier** [sfackler/rust-openssl#1361]
Hello, 

I am trying to cross build from my amd64 to arm32v7 target.
However, I am getting the following issue:

> error: failed to run custom build command for `openssl-sys v0.9.58`

> Caused by:
>   process didn't exit successfully: `/home/rust/src/iotedge/edge-modules/api-proxy-module/target/release/build/openssl-sys-> 9948672ce5c4e726/build-script-main` (exit code: 101)
>   --- stdout
> cargo:rustc-cfg=const_fn
  cargo:rerun-if-env-changed=X86_64_UNKNOWN_LINUX_MUSL_OPENSSL_LIB_DIR
  X86_64_UNKNOWN_LINUX_MUSL_OPENSSL_LIB_DIR unset
  cargo:rerun-if-env-changed=OPENSSL_LIB_DIR
  OPENSSL_LIB_DIR unset
  cargo:rerun-if-env-changed=X86_64_UNKNOWN_LINUX_MUSL_OPENSSL_INCLUDE_DIR
  X86_64_UNKNOWN_LINUX_MUSL_OPENSSL_INCLUDE_DIR unset
  cargo:rerun-if-env-changed=OPENSSL_INCLUDE_DIR
  OPENSSL_INCLUDE_DIR unset
  cargo:rerun-if-env-changed=X86_64_UNKNOWN_LINUX_MUSL_OPENSSL_DIR
  X86_64_UNKNOWN_LINUX_MUSL_OPENSSL_DIR unset
  cargo:rerun-if-env-changed=OPENSSL_DIR
  OPENSSL_DIR unset
  cargo:rerun-if-env-changed=OPENSSL_NO_PKG_CONFIG
  cargo:rerun-if-env-changed=PKG_CONFIG_ALLOW_CROSS_x86_64-unknown-linux-musl
  cargo:rerun-if-env-changed=PKG_CONFIG_ALLOW_CROSS_x86_64_unknown_linux_musl
  cargo:rerun-if-env-changed=TARGET_PKG_CONFIG_ALLOW_CROSS
  cargo:rerun-if-env-changed=PKG_CONFIG_ALLOW_CROSS
  cargo:rerun-if-env-changed=PKG_CONFIG_x86_64-unknown-linux-musl
  cargo:rerun-if-env-changed=PKG_CONFIG_x86_64_unknown_linux_musl
  cargo:rerun-if-env-changed=TARGET_PKG_CONFIG
  cargo:rerun-if-env-changed=PKG_CONFIG
  cargo:rerun-if-env-changed=PKG_CONFIG_SYSROOT_DIR_x86_64-unknown-linux-musl
  cargo:rerun-if-env-changed=PKG_CONFIG_SYSROOT_DIR_x86_64_unknown_linux_musl
  cargo:rerun-if-env-changed=TARGET_PKG_CONFIG_SYSROOT_DIR
  cargo:rerun-if-env-changed=PKG_CONFIG_SYSROOT_DIR
  run pkg_config fail: "pkg-config has not been configured to support cross-compilation.\n\n                Install a sysroot for the target platform and configure it via\n                PKG_CONFIG_SYSROOT_DIR and PKG_CONFIG_PATH, or install a\n                cross-compiling wrapper for pkg-config and set it via\n                PKG_CONFIG environment variable."

Strangely that issue started popping up 6+ months ago.
I can see it is an issue with pkg-config but there are very few source on the internet about it. Does anyone knows what to do?

Thank you!

**Comment by sfackler** [comment]
You need to get a copy of OpenSSL built for arm32v7 and point openssl-sys to it. Alternatively you can enable the `vendored` Cargo feature and have a copy compiled for you and statically linked.

**Comment by huguesBouvier** [comment]
I think there is a problem with the detection:
I have tried both:
1. adding the following
[dependencies]
openssl = { version = "0.10", features = ["vendored"] }
2. Compiling opensll myself:
https://github.com/Azure/iotedge/blob/master/scripts/linux/cross-platform-rust-build.sh#L132
But it still complains about pkg-config not being properly configured.

I see other repo having the same problem: https://travis-ci.org/github/messense/rust-musl-cross
Something has changed but I can't figure what

**Comment by sfackler** [comment]
Looking over the error message in your first post, I'm confused. You said you're targeting arm32v7, but the error message implies that it's targeting x86_64-unknown-linux-musl.

**Comment by huguesBouvier** [comment]
Damn I am sorry, I have many windows opened trying to figure this out, I picked up the wrong one. Sorry about that. Though the error on the main windows was the same.
1. Basically I am using this image to cross compile (https://github.com/emk/rust-musl-builder).
2. This is the command I run:
docker run --rm -it -v "${PROJECT_ROOT}":/home/rust/src ekidd/rust-musl-builder cargo build --target=armv7-unknown-linux-musleabihf --release --manifest-path /home/rust/src/edge-modules/api-proxy-module/Cargo.toml
on that repo:  https://github.com/Azure/iotedge/tree/master/edge-modules/api-proxy-module

After changing "openssl = { version = "0.10", features = ["vendored"] }" I got a different error message:
`error: failed to run custom build command for `openssl-sys v0.9.58`

Caused by:
  process didn't exit successfully: `/home/rust/src/edge-modules/api-proxy-module/target/release/build/openssl-sys-d336d5172efa5f20/build-script-main` (exit code: 101)
  --- stdout
  cargo:rustc-cfg=const_fn
  cargo:rerun-if-env-changed=ARMV7_UNKNOWN_LINUX_MUSLEABIHF_OPENSSL_NO_VENDOR
  ARMV7_UNKNOWN_LINUX_MUSLEABIHF_OPENSSL_NO_VENDOR unset
  cargo:rerun-if-env-changed=OPENSSL_NO_VENDOR
  OPENSSL_NO_VENDOR unset

  --- stderr
  error: could not copy "/home/rust/src/edge-modules/api-proxy-module/target/armv7-unknown-linux-musleabihf/release/build/openssl-sys-954f12af9fa58212/out/probe0.probe0.3a1fbbbh-cgu.0.rcgu.ll" to "/home/rust/src/edge-modules/api-proxy-module/target/armv7-unknown-linux-musleabihf/release/build/openssl-sys-954f12af9fa58212/out/probe0.ll": Operation not permitted (os error 1)

  error: aborting due to previous error

  error: could not copy "/home/rust/src/edge-modules/api-proxy-module/target/armv7-unknown-linux-musleabihf/release/build/openssl-sys-954f12af9fa58212/out/probe1.probe1.3a1fbbbh-cgu.0.rcgu.ll" to "/home/rust/src/edge-modules/api-proxy-module/target/armv7-unknown-linux-musleabihf/release/build/openssl-sys-954f12af9fa58212/out/probe1.ll": Operation not permitted (os error 1)

  error: aborting due to previous error`

Thank you!

**Comment by sfackler** [comment]
It seems like you have some kind of permission errors, maybe due to volume mounts in a Docker setup.

**Comment by huguesBouvier** [comment]
I got around it by using a script instead of dockerfile: https://github.com/Azure/iotedge/blob/master/scripts/linux/cross-platform-rust-build.sh#L101
It compiles! 
I am now getting a similar issue as here: https://github.com/sfackler/rust-openssl/issues/1198

>          /usr/bin/ld: /iotedge/edge-modules/api-proxy-module/target/release/deps/libopenssl_sys-1d369a291d8129e2.rlib(tb_eckey.o): relocation R_X86_64_32 against `.bss' can not be used when making a PIE object; recompile with -fPIC
          /usr/bin/ld: /iotedge/edge-modules/api-proxy-module/target/release/deps/libopenssl_sys-1d369a291d8129e2.rlib(tb_pkmeth.o): relocation R_X86_64_32 against `.bss' can not be used when making a PIE object; recompile with -fPIC
          /usr/bin/ld: /iotedge/edge-modules/api-proxy-module/target/release/deps/libopenssl_sys-1d369a291d8129e2.rlib(err.o): relocation R_X86_64_32 against `.text' can not be used when making a PIE object; recompile with -fPIC
          /usr/bin/ld: /iotedge/edge-modules/api-proxy-module/target/release/deps/libopenssl_sys-1d369a291d8129e2.rlib(digest.o): relocation R_X86_64_32 against `.rodata.str1.1' can not be used when making a PIE object; recompile with -fPIC
          /usr/bin/ld: /iotedge/edge-modules/api-proxy-module/target/release/deps/libopenssl_sys-1d369a291d8129e2.rlib(e_aes.o): relocation R_X86_64_32S against symbol `AES_encrypt' can not be used when making a PIE object; recompile with -fPIC
          /usr/bin/ld: /iotedge/edge-modules/api-proxy-module/target/release/deps/libopenssl_sys-1d369a291d8129e2.rlib(e_aes_cbc_hmac_sha1.o): relocation R_X86_64_32S against

The issue is: If I set openssl to shared then I cannot port the executable to a different target.

**Comment by sfackler** [comment]
Have you tried recompiling OpenSSL with `-fPIC`?

**Comment by huguesBouvier** [comment]
I do use the following script to build openssl (that uses FPIC):

>             cd /tmp && \
            short_version="$(echo "$OPENSSL_VERSION" | sed s\'/[a-z]$//\' )" && \
            curl -fLO "https://www.openssl.org/source/openssl-$OPENSSL_VERSION.tar.gz" || \
                curl -fLO "https://www.openssl.org/source/old/$short_version/openssl-$OPENSSL_VERSION.tar.gz" && \
            tar xvzf "openssl-$OPENSSL_VERSION.tar.gz" && cd "openssl-$OPENSSL_VERSION" && \
            env CC=musl-gcc ./Configure no-shared no-zlib -fPIC --prefix=/usr/local/musl -DOPENSSL_NO_SECURE_MEMORY linux-x86_64 && \
            env C_INCLUDE_PATH=/usr/local/musl/include/ make depend && \
            env C_INCLUDE_PATH=/usr/local/musl/include/ make && \
            sudo make install && \
            sudo rm /usr/local/musl/include/linux /usr/local/musl/include/asm /usr/local/musl/include/asm-generic && \
            rm -r /tmp/*

I have added that "openssl = { version = "0.10", features = ["vendored"] }" maye it has something to do with it? I am new to Rust so I am not very familiar with all the options.

**Comment by sfackler** [comment]
If you have the vendored feature enabled then it shouldn't even be looking at an externally compiled copy of openssl. Can you post a complete, self contained example of a thing that fails to compile?

**Comment by huguesBouvier** [comment]
Thank you for your help! I am closing this, because I got a better understanding of the issue and I don't believe it is a prolem of openssl-sys.
- recently they updated gcc, we need to specify if an hardware is floating point: https://gcc.gnu.org/git/?p=gcc.git&a=commit;h=dff2abcbee65dbb4b7ca3ade0f7622ffdc0af391 and use march=armv7-a+fp instead of march=armv7-a.
- rust-musl-cross deprecated usage of the gcc version that was using floating point by default in january 2020.
- The library that we use to compile had to upgrade their GCC, otherwise they could not use the latest version of rust-musl-cross.

The issue is that rust calls to gcc compiler has not kept up and is still old. It compiles code by using march=armv7-a and not march=armv7-a+fp. That is not clear to me why with this fix I was able to compile: https://github.com/messense/rust-musl-cross/pull/29
